specs/dataclass_transforms.txt

```diff
@@ -0,0 +1,399 @@
+@typing.dataclass_transform(eq_default=True, order_default=True)
+class ModelMeta(type):
+    def __init_subclass__(
+        cls,
+        *,
+        init: bool = True,
+        frozen: bool = False,
+        eq: bool = True,
+        order: bool = True,
+    ):
+        ...
+
+class ModelBase(metaclass=ModelMeta):
+    ...
+
+
+# Example of how this class would be used by code that imports
+# from this library:
+class Customer(ModelBase, init=False, frozen=True, eq=False, order=False):
+    id: int
+    name: str
+```
+
+
+
+Field descriptors
+-----------------
+
+Most libraries that support dataclass-like semantics provide one or more
+"field descriptor" types that allow a class definition to provide additional
+metadata about each field in the class. This metadata can describe, for example,
+default values or indicate whether the field should be included in the
+synthesized `__init__` method.
+
+Field descriptors can be omitted in cases where additional metadata is not
+required or the only metadata is a statically-defined default value.
+
+```python
+@dataclass
+class Employee:
+    # Field with no descriptor
+    name: str
+
+    # Field that uses field descriptor class instance
+    age: Optional[int] = field(default=None, init=False)
+
+    # Field with type annotation and simple initializer to
+    # describe default value
+    is_paid_hourly: bool = True
+
+    # Field with inferred type and simple initializer to
+    # describe default value
+    office_number = "unassigned"
+```
+
+Libraries that support dataclass-like semantics and support field descriptor
+classes typically use common parameter names to construct these field
+descriptors. This specification formalizes the names and meanings of the
+parameters that must be understood for static type checkers. These standardized
+parameters must be keyword-only parameters. Field descriptor classes are
+allowed to use other parameters in their constructors, and those parameters
+can be positional and may use other names.
+
+
+`init` is an optional bool parameter that indicates whether the field should
+be included in the synthesized `__init__` method. If unspecified, it defaults
+to True.
+
+`default` is an optional parameter that provides the default value for the
+field.
+
+`default_factory` is an optional parameter that provides a runtime callback
+that returns the default value for the field. If `default` and `default_value`
+are both unspecified, the field is assumed to have no default value and must be
+provided a value when the class is instantiated.
+
+`alias` is an optional str parameter that provides an alternative name for
+the field. This alternative name is used in the synthesized `__init__` method.
+
+
+This example demonstrates 
+```python
+# Library code (within type stub or inline):
+class ModelField:
+    def __init__(
+        self,
+        *,
+        default: Optional[Any] = ...,
+        init: Optional[bool] = True,
+        **kwargs: Any
+    ) -> None: ...
```

Comment:
> Libraries that support dataclass-like semantics and support field descriptor
classes typically use common parameter names to construct these field
descriptors

so, do that mean libraries that use names not defined in here for field descriptor like types, they need to make breaking changes to opt in? or can they provide decorator to map names?

Reply:
Libraries would need to standardize on these names, which is good for users because the same concepts will have the same names across libraries. This sort of standardization appears to have already happened for the most popular libraries. If a new library is developed that has dataclass-like semantics, they should be highly encouraged to adopt the existing names.

Reply:
looks like this is by design.